### DISCUSSION TITLE
Constraint required default(T) is a valid T

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#4767] [Q&A]
I have a method and I want to specific a constraint on T such that:

T can be a ```struct```
or T can be a ```struct?```
or T can be an ```object?```

Essentially I want to make sure that ```default(T)``` will always result in a valid ```T```` value.

How can I do that?

**Comment by jnm2** [comment]
Can you explain what you mean by "valid `T` value"? `default(T)` is always going to return something that may be null unless you constrain `T : struct`. `default(object)` is `(object?)null`.

```cs
using System;
using System.Diagnostics.CodeAnalysis;

#nullable enable

public class C
{
    public T? M<T>() => default(T);
}
```

**Reply by TonyValenti** [reply]
I want to enable my class to work with structs (nullable or not) and nullable classes.

**Reply by bernd5** [reply]
it works with unconstrained generics. You can write e.g.:

```cs
Something<int> t1;
Something<int?> t2;
Something<object> t3;
Something<object?> t4;

class Something<T>
{
    public T GetDefault(){
        return default(T);
    }
}
```

**Reply by jnm2** [reply]
> I want to enable my class to work with structs (nullable or not) and nullable classes.

@TonyValenti That is exactly what my suggestion of returning `T?` achieves.

**Comment by TonyValenti** [comment]
Maybe I need to clarify:
```
using System;
#nullable enable

namespace ConsoleApp9 {

    public class Foo<T> {
        public T Default() {
            return default(T);
        }
    }

}
```

How do I constrain ```T``` in the above such that default(T) will not produce a warning and I can use my class for ```int```, ```int?``` and ```string?``` but not ```string```?

**Reply by jnm2** [reply]
As in my example, your return type must be `T?`. That's demanded by the nature of what `default(T)` is.

You do not need a constraint for this example. You only need to target C# 9 to be able to use `T?` with unconstrained generic `T`. `T?` on unconstrained `T` means "maybe default."

**Reply by AartBluestoke** [reply]
if you're returning T? does a Foo<int> return Nullable<int>?
otherwise if T is an object type default(T) is null, which will of course warn when returned through a T(non-nullable)

**Reply by Joe4evr** [reply]
> if you're returning T? does a Foo return Nullable?

No.

**Reply by jnm2** [reply]
> No.

And the reason for this is that the runtime doesn't provide this facility.

**Reply by TonyValenti** [reply]
@jnm2 - Oh wow.  Interesting!  I didn't realize that ```T?``` could return a value that cannot contain null.

**Reply by jnm2** [reply]
`T?` means "nullable" unless `T` is not constrained to `class` or `struct`, in which case it means "defaultable."
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/unconstrained-type-parameter-annotations
